### DISCUSSION TITLE
Binary Endpoints for IoT Ecosystem

### DISCUSSION THREAD

**Discussion by oopsmonk** [iotaledger/tips#67] [Selected Ideas]
## Introduction

API endpoints are defined in following TIPs:
* [TIP-25 Node Core REST API](https://github.com/iotaledger/tips/pull/57)
* [TIP-26 UTXO Indexer REST API](https://github.com/iotaledger/tips/pull/62)

Most of the data are communicated via JSON objects which is a human-readable text format. For constrained devices, those text data are not packed and data conversions involve more frequently memory operations, especially allocation and free, which leads to memory fragmentation issues and causes potential system instabilities.

To reduce data conversion and serialization during communications, the node can provide endpoints that communicate with binary payload through HTTP. It has been partially implemented in `/api/v2/messages`. 

The client is able to send a message to the node in JSON string, like

```
curl -X POST http://localhost:14265/api/v2/messages -H 'Content-Type: application/json' -d '{"protocolVersion":2,"parentMessageIds":["1febb3953a740d3d573c0ddedd9bbe646aada4e2db9cf00ec5becc0bc3c01d95","34e87233ad38fb696a9eca00f3f5dc489d8cd4e12f5475bd45739ca24d96535b","62cf0e9db09e3b960486dd392874defa19f0250596fa752a5765b104d55ee53d","ab1f2a6bd28335bae17d0c4f38117189ffaef8397e077a1c2fcc5f2e7dacd8db"],"payload":{"type":0,"essence":{"type":0,"networkId":"8453507715857476362","inputs":[{"type":0,"transactionId":"71cc47ad41681940129eefb6678b5ae1b090aa06283f1a4a6bdae89a394a15a4","transactionOutputIndex":0},{"type":0,"transactionId":"71cc47ad41681940129eefb6678b5ae1b090aa06283f1a4a6bdae89a394a15a4","transactionOutputIndex":1},{"type":0,"transactionId":"71cc47ad41681940129eefb6678b5ae1b090aa06283f1a4a6bdae89a394a15a4","transactionOutputIndex":2},{"type":0,"transactionId":"71cc47ad41681940129eefb6678b5ae1b090aa06283f1a4a6bdae89a394a15a4","transactionOutputIndex":3},{"type":0,"transactionId":"71cc47ad41681940129eefb6678b5ae1b090aa06283f1a4a6bdae89a394a15a4","transactionOutputIndex":4}],"inputsCommitment":"fdd91f8300038ea8371b64b76587d63688cf634ab2159e22574055bd9833cf8e","outputs":[{"type":3,"amount":1000000,"nativeTokens":[],"unlockConditions":[{"type":0,"address":{"type":0,"pubKeyHash":"194eb32b9b6c61207192c7073562a0b3adf50a7c1f268182b552ec8999380acb"}}],"featureBlocks":[]},{"type":3,"amount":1000000,"nativeTokens":[],"unlockConditions":[{"type":0,"address":{"type":0,"pubKeyHash":"194eb32b9b6c61207192c7073562a0b3adf50a7c1f268182b552ec8999380acb"}}],"featureBlocks":[{"type":2,"data":"0d25"}]},{"type":3,"amount":234100,"nativeTokens":[],"unlockConditions":[{"type":0,"address":{"type":0,"pubKeyHash":"194eb32b9b6c61207192c7073562a0b3adf50a7c1f268182b552ec8999380acb"}},{"type":1,"returnAddress":{"type":0,"pubKeyHash":"194eb32b9b6c61207192c7073562a0b3adf50a7c1f268182b552ec8999380acb"},"amount":234000}],"featureBlocks":[]},{"type":3,"amount":1000000,"nativeTokens":[],"unlockConditions":[{"type":0,"address":{"type":0,"pubKeyHash":"194eb32b9b6c61207192c7073562a0b3adf50a7c1f268182b552ec8999380acb"}},{"type":3,"returnAddress":{"type":0,"pubKeyHash":"194eb32b9b6c61207192c7073562a0b3adf50a7c1f268182b552ec8999380acb"},"milestoneIndex":400,"unixTime":0}],"featureBlocks":[]},{"type":3,"amount":1000000,"nativeTokens":[],"unlockConditions":[{"type":0,"address":{"type":0,"pubKeyHash":"194eb32b9b6c61207192c7073562a0b3adf50a7c1f268182b552ec8999380acb"}},{"type":2,"milestoneIndex":400,"unixTime":0}],"featureBlocks":[]}],"payload":null},"unlockBlocks":[{"type":0,"signature":{"type":0,"publicKey":"e364c1734abe4d60507c03a6f34be01fd1a5530680aae95f7ca573afeb300567","signature":"e08ccc6b589b6db27d29216d52c452458cec4ee07ea1e0b8046c8316b560e57a9f593c4018e0888ae62774e91571763358d28e339a63a834f4217fb4dc027a09"}},{"type":1,"reference":0},{"type":1,"reference":0},{"type":1,"reference":0},{"type":1,"reference":0}]},"nonce":"11529215046068502219"}'
```

or binary data, for example

```
curl -X POST http://localhost:14265/api/v2/messages -H 'Content-Type: application/octet-stream' --data-binary @raw
```

The payload size in the example for binary is only 810 bytes, JSON is 2863 bytes, the binary is 3 times smaller. 

Additionally, the following endpoints can be a customized MIME type, instead of `application/octet-stream`, in both requests and responses.

* /api/v2/info
* /api/v2/tips
* /api/v2/outputs/{outputId}
* /api/v2/messages/{messageId}/metadata
* Indexer APIs
* Faucet APIs (Optional)

## Example responses

Excep for data types in [Serialization Primitives](https://github.com/Wollac/protocol-rfcs/blob/serialization/tips/TIP-0021/tip-0021.md), two extended data types are introduced for the Binary Client Response layout:

| Name | Description |
| -------- | -------- |
| (uintX)List(uintY)ByteArray | A dynamically list of ByteArray. A leading `uintX` denotes the length of the list. `uintY` denotes the lenght of ByteArray |
| (uintX)ListByteArray[N]    | A dynamically list of ByteArray[N]. A leading `uintX` denotes the length of the list. |

The value of `X` and `Y` should be one of 8, 16, 32.

### Error response:

```json
{
  "error": {
    "code": 400,
    "message": "invalid data provided"
  }
}
```

Binary:

|    Name | Type | Description |
| --------- | ---------- | -------- |
| Error | uint16   | An error code  |
| Message | (uint32)ByteArray  | Error message with a leading uin32 denotes the string length  |

#### /api/v2/info

JSON

```json
{
  "name": "HORNET",
  "version": "0.6.0-alpha",
  "status": {
    "isHealthy": true,
    "latestMilestoneTimestamp": 1617802102,
    "latestMilestoneIndex": 480,
    "confirmedMilestoneIndex": 480,
    "pruningIndex": 0
  },
  "metrics": {
    "messagesPerSecond": 17,
    "referencedMessagesPerSecond": 16.8,
    "referencedRate": 98.82352941176471
  },
  "protocol": {
    "networkName": "iota-testnet",
    "bech32HRP": "atoi",
    "minPoWScore": 1000,
    "rentStructure": {
      "vByteCost": 500,
      "vByteFactorData": 1,
      "vByteFactorKey": 10
    }
  },
  "features": [
    "PoW"
  ],
  "plugins": [
    "indexer/v1"
  ]
}
```
Binary

| Name | Type | Description |
| -------- | -------- | -------- |
| Error | uint16   | An error code  |
| Name | (uint16)ByteArray   | The name of the node software |
| Version | (uint16)ByteArray   | The version of the node  |
| isHealthy | uint8   | whether the node is healthy  |
| latestMilestoneTimestamp | uint64   | The timestamp of the latest milestone |
| latestMilestoneIndex | uint64   | The latest milestone |
| confirmedMilestoneIndex | uint64   | The index of current confirmed milestone |
| pruningIndex | uint64   | The milestone index since pruning  |
| messagesPerSecond | float(4 bytes)  | The rate of new messages per second  |
| referencedMessagesPerSecond | float(4 byters)   | The rate of referenced messages per second |
| referencedRate | float(4 bytes)   | |
| networkName | (uint16)ByteArray   | The network name of the network  |
| bech32HRP | (uint8)ByteArray   | The HRP prefix for Bech32 addresses  |
| minPoWScore | float(4 bytes)   | The minimum PoW score of the network |
| vByteCost | uint16   | The virtual byte cost |
| vByteFactorData | uint8   | The weight of virtual data  |
| vByteFactorKey | uint8   | The weight of virtual key  |
| features | (uint16)List(uint8)ByteArray  | Enabled features with this node  |
| plugins | (uint16)List(uint8)ByteArray  | Enabled plugin with this node  |

#### /api/v2/tips

JSON

```json
{
  "tipMessageIds": [
    "0xf532a53545103276b46876c473846d98648ee418468bce76df4868648dd73e5d",
    "0x78d546b46aec4557872139a48f66bc567687e8413578a14323548732358914a2"
  ]
}
```

Binary

| Name | Type | Description |
| -------- | -------- | -------- |
| Error | uint16   | An error code  |
| tipMessageIds | (uint8)ListByteArray[32]   | A list of tip message IDs |

#### /api/v2/messages/{messageId}

JSON

```json
{
  "protocolVersion": 1,
  "parentMessageIds": [
    "0xc0b728d4c292860b57747c7bfbd989ccc40be8c76229022644c05e1352bf1ea3",
    "0xe8db39681478cf9826bd2f31d045f16f5e1f478c5abc3030419da7911df9e288"
  ],
  "payload": null,
  "nonce": "1481"
}
```

Binary

| Name | Type | Description |
| -------- | -------- | -------- |
| Error | uint16   | An error code  |
| payload | A serialized message object  | A messge object |

#### /api/v2/outputs/{outputId}

JSON

```json
{
  "messageId": "0x9cd745ef6800c8e8c80b09174ee4b250b3c43dfa62d7c6a4e61f848febf731a0",
  "transactionId": "0x1ee46e19f4219ee65afc10227d0ca22753f76ef32d1e922e5cbe3fbc9b5a5298",
  "outputIndex": 1,
  "isSpent": false,
  "milestoneIndexBooked": 1234567,
  "milestoneTimestampBooked": 1643207146,
  "ledgerIndex": 946704,
  "output": {
    "type": 3,
    "amount": "1000",
    "unlockConditions": [
      {
        "type": 0,
        "address": {
          "type": 0,
          "pubKeyHash": "0x8eaf87ac1f52eb05f2c7c0c15502df990a228838dc37bd18de9503d69afd257d"
        }
      }
    ]
  }
}
```

Binary

| Name | Type | Description |
| -------- | -------- | -------- |
| Error | uint16   | An error code  |
| messageId | ByteArray[32]   | The message ID of the message |
| transactionId | ByteArray[32]  | The transaction ID of this output |
| outputIndex | uint16 | The Index of this output |
| isSpent | uint8  | Whether this output is spent  |
| milestoneIndexBooked | uint64   | The milestone index when this output was booked  |
| milestoneTimestampBooked | uint64  | The milestone timestamp when this output was booked  |
| ledgerIndex | uint64 | The ledger index when this output was available |
| output | A serialized output object | The output object |

#### /api/v2/messages/{messageId}/metadata

JSON

```json
{
  "messageId": "0x2b527cfc71ccb87e986f129116b1d59e28c298400f63dd819f9aa04d4057a06a",
  "parentMessageIds": [
    "0x4547f36e54e6b89ce8ac9d238f0f5fb373fa61c734d8d72f976a370bd76ff647",
    "0x8ca33c5d427a9f2e4850716fdbc48c7653ed27082d64f1637d19dac6d3147072",
    "0xb09ae4f032b9cd52c39c596b3edb9655eabe64bd6de742d6c870c4a7b2c6b0d8",
    "0xe46e92df9cfcc00cc48b807dc9e8e8b01afd6fc79c27446f2061e695b258391e"
  ],
  "isSolid": true,
  "referencedByMilestoneIndex": 54472,
  "ledgerInclusionState": "included"
}
```

Binary

| Name | Type | Description |
| -------- | -------- | -------- |
| Error | uint16   | An error code  |
| messageId | ByteArray[32]   | The message ID of this message  |
| parentMessageIds | (uint8)ListByteArray[32]   | The parent message IDs of this message reference to |
| isSolid | uint8 | Whether this message is solid |
| referencedByMilestoneIndex | uint8 | Whether this message should be promoted (optional) |
| MilestoneIndex | uint64 | If this message is a milestone, this is the milestone index (optional) |
| ledgerInclusionState | (uint8)ByteArray | The ledger inclusion state of the transaction payload (optional) |
| shouldPromote | uint8 | Whether the message should be promoted (optional) |
| shouldReattach | uint8 | Whether the message should be reattached (optional) |
| conflictReason | uint8 | The reason why this message is conflicting (optional) |

### Indexer responses

All Indexer response are using the same response structure, for example

```json
{
  "ledgerIndex": 67633,
  "pageSize": 1000,
  "items": [
    "0x971a55b87617db1bf58f3db7878c19280e4293c836a3966584ea77405d168f5e0000",
    "0x8e9b6a1665db02179e062fdaf8c086f9845c2017b0d9fe70562fc6010d317ece0000"
  ]
}
```
Binary

| Name | Type | Description |
| -------- | -------- | -------- |
| Error | uint16   | An error code  |
| ledgerIndex | uint64 | The ledger index which these outputs available at |
| pageSize | uint32 | the maximum count of results |
| items | (uint16)ListByteArray[34] | A list of outputIDs |
| cursor | (uint16)ByteArray | The cursor to use for getting the next results (optional) |

## Pros

* The data conversion between text and binary can be removed in the client.
* The network bandwidth between clients and nodes is reduced.
* Memory requirement is decreased for the client.

## Cons

* The node needs to maintain both binary and text endpoints
* Binary payload is not easy for debugging

## Conclusion

As constrained clients, such as a system under 100MHz CPU and 256KB memory, the communication should be packed and come with a small memory footprint, this proposal provides a friendly communication channel and opens a door for IoT adoption in the early stage.

**Comment by jyhi** [comment]
A binary format is crucial to resource constrained devices. JSON is compact, but it's not compact enough for resource constrained devices, on which any character-based format would be bloated. By supporting a binary format for APIs on the node software, we allow (even Class 0[^1]) devices to perform actions on the Tangle without hassle.

But by simply saying "a binary format" the structure of data is not defined. Fortunately there has been standard ways to _binary encode_ structured data without losing the structure (which makes it convenient to just convert from and to JSON). Because HTTP is also a character-based protocol, we can also take the transfer protocol into consideration (but not necessarily). I can think of a few possible solutions:

- Data serialization format: **CBOR**, **FlatBuffers**
- Data transfer protocol: **MQTT**/TCP, **CoAP**/UDP

We've had MQTT endpoints running, but with a separate Event API. It's possible to further map and provide REST APIs in MQTT topics to enable more IoT devices to communicate with node software without the bloat of HTTP.

**CBOR**([official website](https://cbor.io/), [Wikipedia](https://en.wikipedia.org/wiki/CBOR), [spec (RFC8949)](https://datatracker.ietf.org/doc/html/rfc8949)) stands for "Concise Binary Object Representation". Its data model is based on JSON, so transformative between JSON and CBOR should be straightforward. It can be sent over any transfer protocol - it's just a way to encode data.

**FlatBuffers** ([official website](https://google.github.io/flatbuffers/index.html), [Wikipedia](https://en.wikipedia.org/wiki/FlatBuffers), [schema compiler](https://github.com/google/flatbuffers)) can be seen as an improvement from [Protocol Buffers](https://en.wikipedia.org/wiki/Protocol_Buffers). It uses schema to define the shape of data, so the serialized binary data has no structure marked in it; in other words, no parsing or unpacking is needed to access the structured data. It's thus more concise than any other serialization format, but it also makes the serialized data non-self-contained.

**CoAP** ([official website](https://coap.technology/), [Wikipedia](https://en.wikipedia.org/wiki/Constrained_Application_Protocol), [spec (RFC7252)](https://datatracker.ietf.org/doc/html/rfc7252/)) is a protocol designed to run on resource constrained devices and UDP. It borrows concepts from HTTP, making common features between the two easy to translate between each other. It's usually seen as a competitor of MQTT, who run (mainly) on TCP and emphasize long connection instead.

One of the ideal combinations could be **CBOR/CoAP**. Both are well-defined, widely-used, widely-supported protocol formats designed specially for resource constrained IoT devices.

It's also fine to just use HTTP to provide the REST APIs. In this way we do [content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation), i.e. checking the `accept:` header sent by a client to determine which serialization format to return. For example, we return CBOR when a request says `accept: application/cbor` we return CBOR; otherwise we return JSON. We don't have to provide a separate endpoint for binary representation (is there any implementation difficulty for this?).

[^1]: _RFC7228 - Terminology for Constrained-Node Networks._ https://datatracker.ietf.org/doc/html/rfc7228

**Reply by oopsmonk** [reply]
> But by simply saying "a binary format" the structure of data is not defined.

Actually, it is defined in protocol models. I've updated some examples in the description.

This proposal is not trying to introduce other serialization tools or network protocols. we can gain the benefits that you mentioned by just changing the content in the HTTP payload.

**Reply by alexsporn** [reply]
> It's also fine to just use HTTP to provide the REST APIs. In this way we do [content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation), i.e. checking the accept: header sent by a client to determine which serialization format to return.

You are right, we don't need the `/api/v2/messages/{messageId}/raw` endpoint as such. in the POST `/api/v2/messages` we do this already using the `Content-Type` header.
But if we do this, then we should define a custom content-type for the IOTA serialisation format and not use `application/octet-stream` just because they are raw bytes.

**Reply by oopsmonk** [reply]
> But if we do this, then we should define a custom content-type for the IOTA serialisation format and not use `application/octet-stream` just because they are raw bytes.

The `octet-stream` subtype is used to indicate that a body contains arbitrary binary data. It perfectly fits to this case.
The `TYPE` and `PADDING` are optional which we don't need. see https://www.rfc-editor.org/rfc/rfc2046.html#section-4.5.1

**Reply by oopsmonk** [reply]
@alexsporn just got my coffee, you are right. a custom MINI type would be a great help for debugging. anyone can write a decoder as needed.

**Reply by arnauorriols** [reply]
> This proposal is not trying to introduce other serialization tools or network protocols. we can gain the benefits that you mentioned by just changing the content in the HTTP payload.

I second @lmy441900 opinion that we should not roll our own adhoc binary format, and instead use any of the already widely supported  ones. As this discussion proposes to extend the binary support to more endpoints of the API, I think it's the right place to open the serialization format debate, if only to discourage moving forward with the proposal as long as it implies further rolling an adhoc binary format. 

> For constrained devices, those text data are not packed and data conversions involve more frequently memory operations, especially allocation and free, which leads to memory fragmentation issues and causes potential system instabilities.

Just for contextualization, in Rust there are options that make parsing JSON a lesser pain in embedded devices. First, [`serde` supports zero-copy deserialization of `&str`](https://serde.rs/lifetimes.html). Secondly, [`serde_json_core`](https://docs.rs/serde-json-core/latest/serde_json_core/) supports deserializing JSON without memory allocation. JSON remains of course a size-inneficient textual format, no question there. 

> It's possible to further map and provide REST APIs in MQTT topics to enable more IoT devices to communicate with node software without the bloat of HTTP.
> [...]
> One of the ideal combinations could be CBOR/CoAP. Both are well-defined, widely-used, widely-supported protocol formats designed specially for resource constrained IoT devices.

In my opinion adding another communication protocol that requires its own server sounds overkill, specially assuming the availability of HTTP/2. I would second mapping the REST endpoints over MQTT topics, although replicating the request/response nature of the REST API over MQTT is not a trivial task _to do right_, and not all constraint devices are suitable for MQTT communication.

> FlatBuffers [...] uses schema to define the shape of data, so the serialized binary data has no structure marked in it; in other words, no parsing or unpacking is needed to access the structured data. It's thus more concise than any other serialization format, but it also makes the serialized data non-self-contained.

I agree with @lmy441900 that CBOR and Flatbuffers are both very well suited options. Considering the coexistence with the JSON format, I would argue in favor of using FlatBuffers instead of CBOR, as the benefits of being a schema-based, zero-copy protocol better set it apart from its JSON counterpart. If schema-less and self-description are considered a hard requirement, [FlexBuffers](https://google.github.io/flatbuffers/flexbuffers.html) can also be a strong contender. The main argument I would have in favour of CBOR is its standardized nature (which is of course not a minor argument). 

If it is (arguably) considered not the time to add any of these binary formats, as said I would discourage moving forward with this proposal and instead wait until a sounder binary support can be considered. Rolling an adhoc binary format in a public API entails a high risk not only to have very low adoption, but actually to be harmful to those 3rd-parties that eventually couple to it.

**Comment by oopsmonk** [comment]
@arnauorriols thanks for your input. this discussion is meant to come out with reasonable support for IoT adoption.

imo, FlatBuffers is not a fit for embedded development, it's designed for gaming and streaming for reducing serialization latency. it bloats the binary data, for example, FlatBuffers needs around 400 bytes of memory to send out 200 bytes of data, the memory consumption is similar to JSON which is a big drawback.

> If it is (arguably) considered not the time to add any of these binary formats, as said I would discourage moving forward with this proposal and instead wait until a sounder binary support can be considered. Rolling an adhoc binary format in a public API entails a high risk not only to have very low adoption, but actually to be harmful to those 3rd-parties that eventually couple to it.

agree, binary communication has low adoption but the protocol shouldn't be changed frequently for a production-ready project.
Could you elaborate more about what does a sounder binary support refer to? might you share an example?

**Comment by oopsmonk** [comment]
Updated Client response objects in the description, thanks for @alexsporn pointing out, the MIMI type can be a customized one and need to be defined instead of `Content-Type: application/octet-stream`. 

discussed with the protocol team, the conclusion is anyone can build an external plugin to connect IoT devices via their own binary format. We don't know the reqs and what they use. We don't want to optimize for constrained devices in Stardust, at least not before the Shimmer launch.

let's move forward and revisit this topic later.
